ASTree.cpp

```diff
@@ -1,3545 +1,3560 @@
+    bool else_pop = false;
+    bool need_try = false;
+    bool variable_annotations = false;
+
+    while (!source.atEof()) {
+#if defined(BLOCK_DEBUG) || defined(STACK_DEBUG)
+        fprintf(stderr, "%-7d", pos);
+    #ifdef STACK_DEBUG
+        fprintf(stderr, "%-5d", (unsigned int)stack_hist.size() + 1);
+    #endif
+    #ifdef BLOCK_DEBUG
+        for (unsigned int i = 0; i < blocks.size(); i++)
+            fprintf(stderr, "    ");
+        fprintf(stderr, "%s (%d)", curblock->type_str(), curblock->end());
+    #endif
+        fprintf(stderr, "\n");
+#endif
+
+        curpos = pos;
+        bc_next(source, mod, opcode, operand, pos);
+
+        if (need_try && opcode != Pyc::SETUP_EXCEPT_A) {
+            need_try = false;
+
+            /* Store the current stack for the except/finally statement(s) */
+            stack_hist.push(stack);
+            PycRef<ASTBlock> tryblock = new ASTBlock(ASTBlock::BLK_TRY, curblock->end(), true);
+            blocks.push(tryblock);
+            curblock = blocks.top();
+        } else if (else_pop
+                && opcode != Pyc::JUMP_FORWARD_A
+                && opcode != Pyc::JUMP_IF_FALSE_A
+                && opcode != Pyc::JUMP_IF_FALSE_OR_POP_A
+                && opcode != Pyc::POP_JUMP_IF_FALSE_A
+                && opcode != Pyc::JUMP_IF_TRUE_A
+                && opcode != Pyc::JUMP_IF_TRUE_OR_POP_A
+                && opcode != Pyc::POP_JUMP_IF_TRUE_A
+                && opcode != Pyc::POP_BLOCK) {
+            else_pop = false;
+
+            PycRef<ASTBlock> prev = curblock;
+            while (prev->end() < pos
+                    && prev->blktype() != ASTBlock::BLK_MAIN) {
+                if (prev->blktype() != ASTBlock::BLK_CONTAINER) {
+                    if (prev->end() == 0) {
+                        break;
+                    }
+
+                    /* We want to keep the stack the same, but we need to pop
+                     * a level off the history. */
+                    //stack = stack_hist.top();
+                    if (!stack_hist.empty())
+                        stack_hist.pop();
+                }
+                blocks.pop();
+
+                if (blocks.empty())
+                    break;
+
+                curblock = blocks.top();
+                curblock->append(prev.cast<ASTNode>());
+
+                prev = curblock;
+
+                CheckIfExpr(stack, curblock);
+            }
+        }
+
+        switch (opcode) {
+        case Pyc::BINARY_OP_A:
+        {
+            ASTBinary::BinOp op = ASTBinary::getBinOpFromOperand(operand);
+            if (op == ASTBinary::BIN_INVALID) {
+                fprintf(stderr, "Unsupported `BINARY_OP` operand value: %s\n", operand);
+            }
+            else {
+                PycRef<ASTNode> right = stack.top();
+                stack.pop();
+                PycRef<ASTNode> left = stack.top();
+                stack.pop();
+                stack.push(new ASTBinary(left, right, op));
+            }

```

Comment:
In order to avoid breaking the stack for the rest of the decompylation, I'd recommend still putting the ASTBinary on the stack even in the failure case...
```suggestion
            ASTBinary::BinOp op = ASTBinary::getBinOpFromOperand(operand);
            if (op == ASTBinary::BIN_INVALID)
                fprintf(stderr, "Unsupported `BINARY_OP` operand value: %s\n", operand);
            PycRef<ASTNode> right = stack.top();
            stack.pop();
            PycRef<ASTNode> left = stack.top();
            stack.pop();
            stack.push(new ASTBinary(left, right, op));
```

This ends up working fine since we still emit a valid `op_str()` even for `BIN_INVALID`

Reply:
True sorry I didn't think about it that's a good idea actually